pkg/plugin/conntrack/_cprog/conntrack.c

```diff
@@ -0,0 +1,245 @@
+    // Check for null parameters.
+    if (!key || !value) {
+        return;
+    }
+    bpf_map_update_elem(&retina_conntrack_map, key, value, BPF_ANY);
+}
+
+/**
+     * Check if a TCP packet should be reported to userspace.
+     * @arg flags The flags of the packet.
+     * @arg value The value of the connection.
+     * Returns true if the packet should be reported to userspace. False otherwise.
+ */
+static __always_inline bool _should_report_tcp_packet(__u8 flags, struct ct_value *value) {
+    // Check for null parameters.
+    if (!value) {
+        return false;
+    }
+
+    __u32 now = bpf_mono_now();
+    __u32 lifetime = value->lifetime;
+    __u8 seen_flags = value->flags_seen;
+    __u32 last_report = value->last_report;
+    // OR the seen flags with the new flags.
+    flags |= seen_flags;
+
+    // Check if the connection timed out or closed.
+    if (now >= lifetime || flags & (TCP_FIN | TCP_RST)) {
+        // The connection is closing or closed. Mark the connection as closing. Update the flags seen and last report time.
+        value->is_closing = 1;
+        value->flags_seen = flags;
+        value->last_report = now;
+        return true; // Report the last packet received.
+    }
+    // Update the lifetime of the connection.
+    value->lifetime = now + CT_CONNECTION_LIFETIME_TCP;
+    // We will only report this packet iff a new flag is seen or the report interval has passed.
+    if (flags != seen_flags || now - last_report >= CT_REPORT_INTERVAL) {
+        value->flags_seen = flags;
+        value->last_report = now;
+        return true;
+    }
+    return false;
+}
+/**
+     * Check if a UDP packet should be reported to userspace.
+     * @arg value The value of the connection.
+     * Returns true if the packet should be reported to userspace. False otherwise.
+ */
+static __always_inline bool _should_report_udp_packet(struct ct_value *value) {
+    // Check for null parameters.
+    if (!value) {
+        return false;
+    }
+    __u32 now = bpf_mono_now();
+    __u32 lifetime = value->lifetime;
+    // Check if the connection timed out.
+    if (now >= lifetime) {
+        return true;
+    }
+    // Update the lifetime of the connection.
+    value->lifetime = now + CT_CONNECTION_LIFETIME_NONTCP;
+    __u32 last_report = value->last_report;
+    // We will only report this packet if the report interval has passed.
+    if (now - last_report >= CT_REPORT_INTERVAL) {
+        value->last_report = now;
+        return true;
+    }
+    return false;
+}
+
+/**
+    * Process a packet and update the connection tracking map.
+    * @arg key The key to be used to lookup the connection in the map.
+    * @arg flags The flags of the packet.
+    * Returns true if the packet should be report to userspace. False otherwise.
+**/
+static __always_inline __attribute__((unused)) bool ct_process_packet(struct ct_v4_key key, __u8 flags) {    
```

Comment:
Write out the algorithm as steps and see how to refactor it in a more manageable way. Currently code complexity (branching) feels too high.